### DISCUSSION TITLE
Change the way Dispatch Output is Implemented and Systems of Neuroprocessors

### DISCUSSION THREAD

**Discussion by keegandent** [TENNLab-UTK/fpga#44] [Ideas]
This occurred to me while we were writing the paper, but I thought it would be silly to try and change during crunch time. Even if everyone needs a break for a bit, I want to jot my thoughts down here anyway.

Right now the Dispatch Sink sends a `num_fires` packet as a header for every timestep. This does not achieve "quiet enough" bandwidth for very sparse output activity. What if instead, we mirror the structure of the Dispatch Source? Let's have "opcodes"! The RUN opcode's operand would translate to "I ran this long without output" and could even be repeated in the event we saturate the operand width. The SPK opcode's operand would represent the index of the neuron that fired on the cycle being described. We pay like two bits for the opcode, but we get potentially orders of magnitude fewer packets depending on the application. I think, for most scenarios, even a RUN 2 or RUN 3 every so often would pay for those bits. The best part is we could size it the exact same way, filling the width in bytes necessary to convey a SPK anyway.

This parallelism starts to reveal something I think would be really cool to explore, potentially for another paper: cooperative networks of hardware neuroprocessors. With some simple switching tables to translate neuron IDs, and some buffering/control to handle synchronization, we could pipeline, fan-out, and fan-in across processors themselves.

There is the question of whether we want to pass along CLR/NOP codes and maybe the latter should be an ERR opcode anyway, as I can't see a use for it. If we need a RUN 0, I'll just rewrite the Dispatch Source to stop assuming RUN 0 means RUN 1. Similarly, for Stream Sink, we could pass back on the CLR bit.

**Comment by jimplank** [comment]
I'm not brain-dead on this yet.  Well, the "systems of neuroprocessors" thing is something I'm not going to think about for a while.  But I do think that there are very simple things that we can do to improve communication:

- Add a command add_periodic_spikes(int neuron_id, int period, int num) that instructs the system to generate periodic spikes on the given input neuron.  I don't know what all that means from an opcode perspective, but it will be super-helpful in a lot of our apps.   Our main encoders are "spikes", which do just that.  To give an example, in the sine function, the encoder always generates periodic spikes on two neurons -- the total spikes between the neurons is 60.  So we trade 60 apply_spikes calls across at least 30 timesteps to two calls on one timestep.

I'll put it on my to-do list to add this to the framework, and then use it in our spike encoders.

- Add an output packet that is something like "skip n", that means that the previous n timesteps had no output.  That way you don't have to have n "0 output spikes" packets.

- There are two processor methods in the framework:

```
    virtual double output_last_fire(int output_id, int network_id = 0) = 0;
    virtual int output_count(int output_id, int network_id = 0) = 0;
```

The whole point of these was so that neuroprocessors don't have to output spikes, but instead output this information, both of which cost an integer.  That way you don't have to have any output except after n timesteps in a RUN N call.  DBSCAN is the only application where we look at individual spikes.  All of the others use spike decoders which use counts or last spike times (TTLS).

I think these things should be simple to implement and don't cost much additional memory (as opposed to buffering, which does).

**Reply by keegandent** [reply]
My only concern is adding a different periodic spike command to the dispatch opcodes requires specifying another processor parameter for exactly how high that number could be, and I worry if we're not opening Pandora's Box for processor instructions...

I've started designing a network on paper that could take two consecutive spikes on a single input neuron and repeat the period ad infinitum, or use an array of input neurons to encode a period value. I'm struggling to do this in a way that is agnostic to leak mode, so let me know which one I should focus on first and which methodology you like more (though I think one is just a "rotation" of the other). A clearing neuron to turn off periodic spiking and allow a new period would be simple.

**Comment by jimplank** [comment]
That's an interesting option, and I think it would be cool to test one vs another.  However, I think it's worth the potential bloat to add the instruction, and to have its max value be a proc param.  If we want to demonstrate performance, we have at least a 1000 to 1 communication to neuroprocessor-bandwidth gap to make up, and this instruction can do that quite easily, so I feel like it's worth the pain.  I'll put it into the framework and into RISP (next week probably).  Happy to have more discussion — Jim

On Oct 22, 2024, at 3:46 PM, Keegan Dent ***@***.***> wrote:



My only concern is adding a different periodic spike command to the dispatch opcodes requires specifying another processor parameter for exactly how high that number could be, and I worry if we're not opening Pandora's Box for processor instructions...

I've started designing a network on paper that could take two consecutive spikes on a single input neuron and repeat the period ad infinitum, or use an array of input neurons to encode a period value. I'm struggling to do this in a way that is agnostic to leak mode, so let me know which one I should focus on first and which methodology you like more (though I think one is just a "rotation" of the other). A clearing neuron to turn off periodic spiking and allow a new period would be simple.

**Comment by keegandent** [comment]
I think between the three of us, we have delivered or at least are in the process of delivering the ideas mentioned here.
